lib/spa-deploy/spa-deploy-construct.ts

```diff
@@ -168,6 +171,7 @@ export class SPADeploy extends cdk.Construct {
 
       new s3deploy.BucketDeployment(this, 'BucketDeployment', {
         sources: [s3deploy.Source.asset(config.websiteFolder)],
+        role: config.role,

```

Comment:
what happens if you don't pass in a role? Will this not set it to undefined? Not sure how Cloudformation would handle that

Reply:
In the instance that it is undefined, a role is automatically created by the `BucketDeployment` L2 construct as per [here](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts#L86-L91).

Reply:
right, but what I mean is this will pass the JSON property "role" in with a value of undefined. Which is different to not passing in a role. So if CF simply checks for the existence of the property then this would fail to deploy if you know what I mean

Reply:
Fair question.

Call chain is:

1. [`bucket-deployment.ts` `196` calls `SingletonFunction` constructor](https://github.com/aws/aws-cdk/blob/76f06fc2ae4404242e5854d2eeaf4f11b98f98f0/packages/%40aws-cdk/aws-s3-deployment/lib/bucket-deployment.ts#L196-L208).
1. `SingletonFunction` constructor, calls private function [`ensureLambda`](https://github.com/aws/aws-cdk/blob/76f06fc2ae4404242e5854d2eeaf4f11b98f98f0/packages/%40aws-cdk/aws-lambda/lib/singleton-lambda.ts#L56)
1. [`ensureLambda` function calls `LambdaFunction` constructor.](https://github.com/aws/aws-cdk/blob/76f06fc2ae4404242e5854d2eeaf4f11b98f98f0/packages/%40aws-cdk/aws-lambda/lib/singleton-lambda.ts#L134)
1. Constructor of `LambdaFunction` [checks that `props.role` is defined, if not it creates a new role on line `580`](https://github.com/aws/aws-cdk/blob/76f06fc2ae4404242e5854d2eeaf4f11b98f98f0/packages/%40aws-cdk/aws-lambda/lib/function.ts#L580-L583)
1. `this.role` in `SingletonFunction` constructor [sets against result of call to `ensureLambda`](https://github.com/aws/aws-cdk/blob/76f06fc2ae4404242e5854d2eeaf4f11b98f98f0/packages/%40aws-cdk/aws-lambda/lib/singleton-lambda.ts#L61)

Have added a test there to verify the role gets auto created if explicitly set to `undefined`